{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "cdd0a298-de34-4915-8bae-192d11cb06ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f8cf0b5f-e0cb-4ac6-b6d0-aee02c4966a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file_path = \"/data/projects/Resources/Gencode_genome_annotation/gencode.v40.annotation.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8520ea38-2cb3-4d98-bd36-8f8a77f59bd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_attributes(attribute_string):\n",
    "    # Split each attribute into key and value, then strip quotes and spaces\n",
    "    return dict(item.strip().replace('\"', '').split(' ') for item in attribute_string if item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "de06ed69-923c-4f70-8141-686c0008151f",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_names = [\n",
    "    \"seqname\", \"source\", \"feature\", \"start\", \"end\",\n",
    "    \"score\", \"strand\", \"frame\", \"attribute\"\n",
    "]\n",
    "\n",
    "# Read the GTF file\n",
    "gtf_df = pd.read_csv(gtf_file_path, sep=\"\\t\", comment='#', header=None, names=column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c42d403f-4324-4b16-b24b-c69aba6bf33c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Split the 'attribute' string into a list of strings for each key-value pair\n",
    "attributes_list = gtf_df['attribute'].str.split(';')\n",
    "\n",
    "\n",
    "# Apply the function to each row's attribute list\n",
    "attributes_dicts = attributes_list.apply(parse_attributes)\n",
    "\n",
    "# Step 3: Convert the list of dictionaries into a DataFrame\n",
    "attributes_df = pd.DataFrame(list(attributes_dicts))\n",
    "\n",
    "# Step 4: Combine the new attributes DataFrame with the original gtf_df\n",
    "# This step assumes that the indexes are aligned and can be directly concatenated\n",
    "combined_df = pd.concat([gtf_df, attributes_df], axis=1)\n",
    "\n",
    "# Optionally, you can drop the original 'attribute' column if it's no longer needed\n",
    "combined_df = combined_df.drop('attribute', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8d716ea9-00ee-4cc9-a821-43e4152abb5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283857</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283858</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283859</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283860</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283861</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3283862 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature  start    end score strand frame  \\\n",
       "0          chr1   HAVANA        gene  11869  14409     .      +     .   \n",
       "1          chr1   HAVANA  transcript  11869  14409     .      +     .   \n",
       "2          chr1   HAVANA        exon  11869  12227     .      +     .   \n",
       "3          chr1   HAVANA        exon  12613  12721     .      +     .   \n",
       "4          chr1   HAVANA        exon  13221  14409     .      +     .   \n",
       "...         ...      ...         ...    ...    ...   ...    ...   ...   \n",
       "3283857    chrM  ENSEMBL  transcript  15888  15953     .      +     .   \n",
       "3283858    chrM  ENSEMBL        exon  15888  15953     .      +     .   \n",
       "3283859    chrM  ENSEMBL        gene  15956  16023     .      -     .   \n",
       "3283860    chrM  ENSEMBL  transcript  15956  16023     .      -     .   \n",
       "3283861    chrM  ENSEMBL        exon  15956  16023     .      -     .   \n",
       "\n",
       "                   gene_id                           gene_type gene_name  \\\n",
       "0        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "1        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "2        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "3        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "4        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "...                    ...                                 ...       ...   \n",
       "3283857  ENSG00000210195.2                             Mt_tRNA     MT-TT   \n",
       "3283858  ENSG00000210195.2                             Mt_tRNA     MT-TT   \n",
       "3283859  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "3283860  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "3283861  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "\n",
       "        level     hgnc_id           havana_gene      transcript_id  \\\n",
       "0           2  HGNC:37102  OTTHUMG00000000961.2                NaN   \n",
       "1           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "...       ...         ...                   ...                ...   \n",
       "3283857     3   HGNC:7499                   NaN  ENST00000387460.2   \n",
       "3283858     3   HGNC:7499                   NaN  ENST00000387460.2   \n",
       "3283859     3   HGNC:7494                   NaN                NaN   \n",
       "3283860     3   HGNC:7494                   NaN  ENST00000387461.2   \n",
       "3283861     3   HGNC:7494                   NaN  ENST00000387461.2   \n",
       "\n",
       "              transcript_type transcript_name transcript_support_level  \\\n",
       "0                         NaN             NaN                      NaN   \n",
       "1        processed_transcript     DDX11L1-202                        1   \n",
       "2        processed_transcript     DDX11L1-202                        1   \n",
       "3        processed_transcript     DDX11L1-202                        1   \n",
       "4        processed_transcript     DDX11L1-202                        1   \n",
       "...                       ...             ...                      ...   \n",
       "3283857               Mt_tRNA       MT-TT-201                       NA   \n",
       "3283858               Mt_tRNA       MT-TT-201                       NA   \n",
       "3283859                   NaN             NaN                      NaN   \n",
       "3283860               Mt_tRNA       MT-TP-201                       NA   \n",
       "3283861               Mt_tRNA       MT-TP-201                       NA   \n",
       "\n",
       "                       tag     havana_transcript exon_number  \\\n",
       "0                      NaN                   NaN         NaN   \n",
       "1                    basic  OTTHUMT00000362751.1         NaN   \n",
       "2                    basic  OTTHUMT00000362751.1           1   \n",
       "3                    basic  OTTHUMT00000362751.1           2   \n",
       "4                    basic  OTTHUMT00000362751.1           3   \n",
       "...                    ...                   ...         ...   \n",
       "3283857  Ensembl_canonical                   NaN         NaN   \n",
       "3283858  Ensembl_canonical                   NaN           1   \n",
       "3283859                NaN                   NaN         NaN   \n",
       "3283860  Ensembl_canonical                   NaN         NaN   \n",
       "3283861  Ensembl_canonical                   NaN           1   \n",
       "\n",
       "                   exon_id  ont protein_id ccdsid  \n",
       "0                      NaN  NaN        NaN    NaN  \n",
       "1                      NaN  NaN        NaN    NaN  \n",
       "2        ENSE00002234944.1  NaN        NaN    NaN  \n",
       "3        ENSE00003582793.1  NaN        NaN    NaN  \n",
       "4        ENSE00002312635.1  NaN        NaN    NaN  \n",
       "...                    ...  ...        ...    ...  \n",
       "3283857                NaN  NaN        NaN    NaN  \n",
       "3283858  ENSE00001544475.2  NaN        NaN    NaN  \n",
       "3283859                NaN  NaN        NaN    NaN  \n",
       "3283860                NaN  NaN        NaN    NaN  \n",
       "3283861  ENSE00001544473.2  NaN        NaN    NaN  \n",
       "\n",
       "[3283862 rows x 25 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "edd0e18a-cfdc-4b88-b467-17907cf847bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "transcript_type\n",
       "IG_C_gene                                 267\n",
       "IG_C_pseudogene                            24\n",
       "IG_D_gene                                 115\n",
       "IG_J_gene                                  58\n",
       "IG_J_pseudogene                             6\n",
       "IG_V_gene                                 989\n",
       "IG_V_pseudogene                           474\n",
       "IG_pseudogene                               2\n",
       "Mt_rRNA                                     4\n",
       "Mt_tRNA                                    44\n",
       "TEC                                      2303\n",
       "TR_C_gene                                  64\n",
       "TR_D_gene                                  12\n",
       "TR_J_gene                                 237\n",
       "TR_J_pseudogene                             8\n",
       "TR_V_gene                                 717\n",
       "TR_V_pseudogene                            90\n",
       "lncRNA                                 237669\n",
       "miRNA                                    3758\n",
       "misc_RNA                                 4424\n",
       "non_stop_decay                           1657\n",
       "nonsense_mediated_decay                516491\n",
       "polymorphic_pseudogene                   1334\n",
       "processed_pseudogene                    21933\n",
       "processed_transcript                   172709\n",
       "protein_coding                        2028295\n",
       "pseudogene                                 49\n",
       "rRNA                                       94\n",
       "rRNA_pseudogene                           994\n",
       "retained_intron                        201411\n",
       "ribozyme                                   16\n",
       "sRNA                                       10\n",
       "scRNA                                       2\n",
       "scaRNA                                     98\n",
       "snRNA                                    3802\n",
       "snoRNA                                   1886\n",
       "transcribed_processed_pseudogene         1181\n",
       "transcribed_unitary_pseudogene           1448\n",
       "transcribed_unprocessed_pseudogene       6822\n",
       "translated_processed_pseudogene             4\n",
       "translated_unprocessed_pseudogene          18\n",
       "unitary_pseudogene                        371\n",
       "unprocessed_pseudogene                  10426\n",
       "vault_RNA                                   2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.groupby(\"transcript_type\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9c58a4a7-3fe6-4410-95a2-1074fdcd76cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv(\"/data/projects/Resources/Gencode_genome_annotation/gencode_annotation_V40.tsv\", sep= \"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b8d86e06-cf45-499a-9fd9-2e2a8178050b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283698</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57209306</td>\n",
       "      <td>57210051</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483079.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-210</td>\n",
       "      <td>1</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058833.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283701</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57209887</td>\n",
       "      <td>57212186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000496301.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-215</td>\n",
       "      <td>2</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058827.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283704</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57210344</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483286.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-211</td>\n",
       "      <td>1</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058834.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283708</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57210591</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000464205.6_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>WASH6P-205</td>\n",
       "      <td>2</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058835.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283713</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246587 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature     start       end score strand frame  \\\n",
       "1          chr1   HAVANA  transcript     11869     14409     .      +     .   \n",
       "5          chr1   HAVANA  transcript     12010     13670     .      +     .   \n",
       "13         chr1   HAVANA  transcript     14404     29570     .      -     .   \n",
       "26         chr1  ENSEMBL  transcript     17369     17436     .      -     .   \n",
       "29         chr1   HAVANA  transcript     29554     31097     .      +     .   \n",
       "...         ...      ...         ...       ...       ...   ...    ...   ...   \n",
       "3283698    chrY   HAVANA  transcript  57209306  57210051     .      +     .   \n",
       "3283701    chrY   HAVANA  transcript  57209887  57212186     .      +     .   \n",
       "3283704    chrY   HAVANA  transcript  57210344  57212074     .      +     .   \n",
       "3283708    chrY   HAVANA  transcript  57210591  57212074     .      +     .   \n",
       "3283713    chrY   HAVANA  transcript  57212184  57214397     .      -     .   \n",
       "\n",
       "                          gene_id                           gene_type  \\\n",
       "1               ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "5               ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "13              ENSG00000227232.5              unprocessed_pseudogene   \n",
       "26              ENSG00000278267.1                               miRNA   \n",
       "29              ENSG00000243485.5                              lncRNA   \n",
       "...                           ...                                 ...   \n",
       "3283698  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283701  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283704  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283708  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283713   ENSG00000227159.8_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "           gene_name level     hgnc_id           havana_gene  \\\n",
       "1            DDX11L1     2  HGNC:37102  OTTHUMG00000000961.2   \n",
       "5            DDX11L1     2  HGNC:37102  OTTHUMG00000000961.2   \n",
       "13            WASH7P     2  HGNC:38034  OTTHUMG00000000958.1   \n",
       "26         MIR6859-1     3  HGNC:50039                   NaN   \n",
       "29       MIR1302-2HG     2  HGNC:52482  OTTHUMG00000000959.2   \n",
       "...              ...   ...         ...                   ...   \n",
       "3283698       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283701       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283704       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283708       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283713     DDX11L16     2  HGNC:37115  OTTHUMG00000022678.1   \n",
       "\n",
       "                   transcript_id                     transcript_type  \\\n",
       "1              ENST00000456328.2                processed_transcript   \n",
       "5              ENST00000450305.2  transcribed_unprocessed_pseudogene   \n",
       "13             ENST00000488147.1              unprocessed_pseudogene   \n",
       "26             ENST00000619216.1                               miRNA   \n",
       "29             ENST00000473358.1                              lncRNA   \n",
       "...                          ...                                 ...   \n",
       "3283698  ENST00000483079.6_PAR_Y                     retained_intron   \n",
       "3283701  ENST00000496301.6_PAR_Y                     retained_intron   \n",
       "3283704  ENST00000483286.6_PAR_Y                     retained_intron   \n",
       "3283708  ENST00000464205.6_PAR_Y                processed_transcript   \n",
       "3283713  ENST00000507418.6_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "         transcript_name transcript_support_level                tag  \\\n",
       "1            DDX11L1-202                        1              basic   \n",
       "5            DDX11L1-201                       NA  Ensembl_canonical   \n",
       "13            WASH7P-201                       NA  Ensembl_canonical   \n",
       "26         MIR6859-1-201                       NA  Ensembl_canonical   \n",
       "29       MIR1302-2HG-202                        5  Ensembl_canonical   \n",
       "...                  ...                      ...                ...   \n",
       "3283698       WASH6P-210                        1                PAR   \n",
       "3283701       WASH6P-215                        2                PAR   \n",
       "3283704       WASH6P-211                        1                PAR   \n",
       "3283708       WASH6P-205                        2                PAR   \n",
       "3283713     DDX11L16-201                       NA                PAR   \n",
       "\n",
       "            havana_transcript exon_number exon_id          ont protein_id  \\\n",
       "1        OTTHUMT00000362751.1         NaN     NaN          NaN        NaN   \n",
       "5        OTTHUMT00000002844.2         NaN     NaN  PGO:0000019        NaN   \n",
       "13       OTTHUMT00000002839.1         NaN     NaN  PGO:0000005        NaN   \n",
       "26                        NaN         NaN     NaN          NaN        NaN   \n",
       "29       OTTHUMT00000002840.1         NaN     NaN          NaN        NaN   \n",
       "...                       ...         ...     ...          ...        ...   \n",
       "3283698  OTTHUMT00000058833.1         NaN     NaN          NaN        NaN   \n",
       "3283701  OTTHUMT00000058827.1         NaN     NaN          NaN        NaN   \n",
       "3283704  OTTHUMT00000058834.1         NaN     NaN          NaN        NaN   \n",
       "3283708  OTTHUMT00000058835.1         NaN     NaN          NaN        NaN   \n",
       "3283713  OTTHUMT00000058841.1         NaN     NaN  PGO:0000005        NaN   \n",
       "\n",
       "        ccdsid  \n",
       "1          NaN  \n",
       "5          NaN  \n",
       "13         NaN  \n",
       "26         NaN  \n",
       "29         NaN  \n",
       "...        ...  \n",
       "3283698    NaN  \n",
       "3283701    NaN  \n",
       "3283704    NaN  \n",
       "3283708    NaN  \n",
       "3283713    NaN  \n",
       "\n",
       "[246587 rows x 25 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter rows where 'feature' is 'transcript' and 'seqname' is not 'chrM'\n",
    "gtf_transcript = combined_df[(combined_df[\"feature\"] == \"transcript\") & (combined_df[\"seqname\"] != \"chrM\")]\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "gtf_transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "11332d68-96ff-4430-acd2-099cd9ca767c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_tss_regions_from_df(tss_df):\n",
    "    \"\"\"\n",
    "    Extracts TSS (+50 and -50 bp) regions from the GTF DataFrame.\n",
    "    :param df: A pandas DataFrame containing GTF data.\n",
    "    :return: A DataFrame with TSS regions.\n",
    "    \"\"\"\n",
    "\n",
    "    # Calculate TSS based on strand\n",
    "    tss_df['TSS'] = tss_df.apply(lambda x: x['start'] if x['strand'] == '+' else x['end'], axis=1)\n",
    "\n",
    "    # Adjust TSS based on strand for +50 and -50 bp\n",
    "    tss_df['start_adj'] = tss_df.apply(lambda x: x['TSS'] - 50 if x['strand'] == '+' else x['TSS'], axis=1)\n",
    "    tss_df['end_adj'] = tss_df.apply(lambda x: x['TSS'] + 50 if x['strand'] == '+' else x['TSS'] + 100, axis=1)\n",
    "\n",
    "    # Select and rename relevant columns\n",
    "    tss_regions = tss_df[['seqname', 'TSS','start_adj', 'end_adj', 'strand', 'gene_id', 'gene_type', 'gene_name', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name']]\n",
    "    tss_regions.columns = ['Chromosome', 'TSS', '50BP_Start', '50BP_End', 'Strand', 'GeneID', 'gene_type', 'gene_name', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name']\n",
    "\n",
    "    return tss_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ee7903b9-6683-4c34-a1e2-1c00004cba19",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_103852/3899473439.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tss_df['TSS'] = tss_df.apply(lambda x: x['start'] if x['strand'] == '+' else x['end'], axis=1)\n",
      "/tmp/ipykernel_103852/3899473439.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tss_df['start_adj'] = tss_df.apply(lambda x: x['TSS'] - 50 if x['strand'] == '+' else x['TSS'], axis=1)\n",
      "/tmp/ipykernel_103852/3899473439.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tss_df['end_adj'] = tss_df.apply(lambda x: x['TSS'] + 50 if x['strand'] == '+' else x['TSS'] + 100, axis=1)\n"
     ]
    }
   ],
   "source": [
    "tss_regions_df = extract_tss_regions_from_df(gtf_transcript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "93ef4cae-5e4b-4c46-8c07-d0da487fff46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>TSS</th>\n",
       "      <th>50BP_Start</th>\n",
       "      <th>50BP_End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>11819</td>\n",
       "      <td>11919</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12010</td>\n",
       "      <td>11960</td>\n",
       "      <td>12060</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29570</td>\n",
       "      <td>29570</td>\n",
       "      <td>29670</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17436</td>\n",
       "      <td>17436</td>\n",
       "      <td>17536</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>29504</td>\n",
       "      <td>29604</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283698</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209306</td>\n",
       "      <td>57209256</td>\n",
       "      <td>57209356</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483079.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283701</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209887</td>\n",
       "      <td>57209837</td>\n",
       "      <td>57209937</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000496301.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283704</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210344</td>\n",
       "      <td>57210294</td>\n",
       "      <td>57210394</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483286.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283708</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210591</td>\n",
       "      <td>57210541</td>\n",
       "      <td>57210641</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000464205.6_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>WASH6P-205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283713</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57214397</td>\n",
       "      <td>57214397</td>\n",
       "      <td>57214497</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246587 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome       TSS  50BP_Start  50BP_End Strand  \\\n",
       "1             chr1     11869       11819     11919      +   \n",
       "5             chr1     12010       11960     12060      +   \n",
       "13            chr1     29570       29570     29670      -   \n",
       "26            chr1     17436       17436     17536      -   \n",
       "29            chr1     29554       29504     29604      +   \n",
       "...            ...       ...         ...       ...    ...   \n",
       "3283698       chrY  57209306    57209256  57209356      +   \n",
       "3283701       chrY  57209887    57209837  57209937      +   \n",
       "3283704       chrY  57210344    57210294  57210394      +   \n",
       "3283708       chrY  57210591    57210541  57210641      +   \n",
       "3283713       chrY  57214397    57214397  57214497      -   \n",
       "\n",
       "                           GeneID                           gene_type  \\\n",
       "1               ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "5               ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "13              ENSG00000227232.5              unprocessed_pseudogene   \n",
       "26              ENSG00000278267.1                               miRNA   \n",
       "29              ENSG00000243485.5                              lncRNA   \n",
       "...                           ...                                 ...   \n",
       "3283698  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283701  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283704  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283708  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283713   ENSG00000227159.8_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "           gene_name     hgnc_id           havana_gene  \\\n",
       "1            DDX11L1  HGNC:37102  OTTHUMG00000000961.2   \n",
       "5            DDX11L1  HGNC:37102  OTTHUMG00000000961.2   \n",
       "13            WASH7P  HGNC:38034  OTTHUMG00000000958.1   \n",
       "26         MIR6859-1  HGNC:50039                   NaN   \n",
       "29       MIR1302-2HG  HGNC:52482  OTTHUMG00000000959.2   \n",
       "...              ...         ...                   ...   \n",
       "3283698       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283701       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283704       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283708       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283713     DDX11L16  HGNC:37115  OTTHUMG00000022678.1   \n",
       "\n",
       "                   transcript_id                     transcript_type  \\\n",
       "1              ENST00000456328.2                processed_transcript   \n",
       "5              ENST00000450305.2  transcribed_unprocessed_pseudogene   \n",
       "13             ENST00000488147.1              unprocessed_pseudogene   \n",
       "26             ENST00000619216.1                               miRNA   \n",
       "29             ENST00000473358.1                              lncRNA   \n",
       "...                          ...                                 ...   \n",
       "3283698  ENST00000483079.6_PAR_Y                     retained_intron   \n",
       "3283701  ENST00000496301.6_PAR_Y                     retained_intron   \n",
       "3283704  ENST00000483286.6_PAR_Y                     retained_intron   \n",
       "3283708  ENST00000464205.6_PAR_Y                processed_transcript   \n",
       "3283713  ENST00000507418.6_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "         transcript_name  \n",
       "1            DDX11L1-202  \n",
       "5            DDX11L1-201  \n",
       "13            WASH7P-201  \n",
       "26         MIR6859-1-201  \n",
       "29       MIR1302-2HG-202  \n",
       "...                  ...  \n",
       "3283698       WASH6P-210  \n",
       "3283701       WASH6P-215  \n",
       "3283704       WASH6P-211  \n",
       "3283708       WASH6P-205  \n",
       "3283713     DDX11L16-201  \n",
       "\n",
       "[246587 rows x 13 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_regions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8e7bb847-f105-4254-b83b-cb99ba732a05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a custom aggregation function\n",
    "def custom_agg(series):\n",
    "    # If all values are the same, return any one of them; otherwise, return the list of unique values\n",
    "    if series.nunique() == 1:\n",
    "        return series.iloc[0]\n",
    "    else:\n",
    "        return list(series.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "08c15241-179b-4766-ac9b-5a46442a087f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group by the columns 'Chromosome', 'Start', 'End', 'Strand'\n",
    "# and aggregate other columns\n",
    "unique_tss_df = tss_regions_df.groupby(['Chromosome', '50BP_Start', '50BP_End', 'Strand']).agg(custom_agg).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a2158629-241a-46e0-ba4d-f582f5143111",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>50BP_Start</th>\n",
       "      <th>50BP_End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11819</td>\n",
       "      <td>11919</td>\n",
       "      <td>+</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11960</td>\n",
       "      <td>12060</td>\n",
       "      <td>+</td>\n",
       "      <td>12010</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17436</td>\n",
       "      <td>17536</td>\n",
       "      <td>-</td>\n",
       "      <td>17436</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29504</td>\n",
       "      <td>29604</td>\n",
       "      <td>+</td>\n",
       "      <td>29554</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29570</td>\n",
       "      <td>29670</td>\n",
       "      <td>-</td>\n",
       "      <td>29570</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215293</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209256</td>\n",
       "      <td>57209356</td>\n",
       "      <td>+</td>\n",
       "      <td>57209306</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483079.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215294</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209837</td>\n",
       "      <td>57209937</td>\n",
       "      <td>+</td>\n",
       "      <td>57209887</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000496301.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215295</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210294</td>\n",
       "      <td>57210394</td>\n",
       "      <td>+</td>\n",
       "      <td>57210344</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483286.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215296</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210541</td>\n",
       "      <td>57210641</td>\n",
       "      <td>+</td>\n",
       "      <td>57210591</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000464205.6_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>WASH6P-205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215297</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57214397</td>\n",
       "      <td>57214497</td>\n",
       "      <td>-</td>\n",
       "      <td>57214397</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215298 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  50BP_Start  50BP_End Strand       TSS  \\\n",
       "0            chr1       11819     11919      +     11869   \n",
       "1            chr1       11960     12060      +     12010   \n",
       "2            chr1       17436     17536      -     17436   \n",
       "3            chr1       29504     29604      +     29554   \n",
       "4            chr1       29570     29670      -     29570   \n",
       "...           ...         ...       ...    ...       ...   \n",
       "215293       chrY    57209256  57209356      +  57209306   \n",
       "215294       chrY    57209837  57209937      +  57209887   \n",
       "215295       chrY    57210294  57210394      +  57210344   \n",
       "215296       chrY    57210541  57210641      +  57210591   \n",
       "215297       chrY    57214397  57214497      -  57214397   \n",
       "\n",
       "                          GeneID                           gene_type  \\\n",
       "0              ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "1              ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "2              ENSG00000278267.1                               miRNA   \n",
       "3              ENSG00000243485.5                              lncRNA   \n",
       "4              ENSG00000227232.5              unprocessed_pseudogene   \n",
       "...                          ...                                 ...   \n",
       "215293  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215294  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215295  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215296  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215297   ENSG00000227159.8_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "          gene_name     hgnc_id           havana_gene  \\\n",
       "0           DDX11L1  HGNC:37102  OTTHUMG00000000961.2   \n",
       "1           DDX11L1  HGNC:37102  OTTHUMG00000000961.2   \n",
       "2         MIR6859-1  HGNC:50039                 [nan]   \n",
       "3       MIR1302-2HG  HGNC:52482  OTTHUMG00000000959.2   \n",
       "4            WASH7P  HGNC:38034  OTTHUMG00000000958.1   \n",
       "...             ...         ...                   ...   \n",
       "215293       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215294       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215295       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215296       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215297     DDX11L16  HGNC:37115  OTTHUMG00000022678.1   \n",
       "\n",
       "                  transcript_id                     transcript_type  \\\n",
       "0             ENST00000456328.2                processed_transcript   \n",
       "1             ENST00000450305.2  transcribed_unprocessed_pseudogene   \n",
       "2             ENST00000619216.1                               miRNA   \n",
       "3             ENST00000473358.1                              lncRNA   \n",
       "4             ENST00000488147.1              unprocessed_pseudogene   \n",
       "...                         ...                                 ...   \n",
       "215293  ENST00000483079.6_PAR_Y                     retained_intron   \n",
       "215294  ENST00000496301.6_PAR_Y                     retained_intron   \n",
       "215295  ENST00000483286.6_PAR_Y                     retained_intron   \n",
       "215296  ENST00000464205.6_PAR_Y                processed_transcript   \n",
       "215297  ENST00000507418.6_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "        transcript_name  \n",
       "0           DDX11L1-202  \n",
       "1           DDX11L1-201  \n",
       "2         MIR6859-1-201  \n",
       "3       MIR1302-2HG-202  \n",
       "4            WASH7P-201  \n",
       "...                 ...  \n",
       "215293       WASH6P-210  \n",
       "215294       WASH6P-215  \n",
       "215295       WASH6P-211  \n",
       "215296       WASH6P-205  \n",
       "215297     DDX11L16-201  \n",
       "\n",
       "[215298 rows x 13 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "087a9752-9305-4896-b681-00170e8b0f49",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_genome(fasta_file):\n",
    "    \"\"\"\n",
    "    Reads a genome fasta file and returns a dictionary of chromosomes and sequences.\n",
    "    :param fasta_file: Path to the genome fasta file.\n",
    "    :return: A dictionary with chromosome names as keys and sequences as values.\n",
    "    \"\"\"\n",
    "    genome = {}\n",
    "    for seq_record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        genome[seq_record.id] = seq_record.seq\n",
    "    return genome\n",
    "\n",
    "def fetch_sequences_for_regions(df, genome):\n",
    "    \"\"\"\n",
    "    Fetches sequences for regions specified in the DataFrame from the genome.\n",
    "    :param df: DataFrame with the regions (must contain 'Chromosome', 'Start', 'End', and 'Strand' columns).\n",
    "    :param genome: Dictionary of genome sequences.\n",
    "    :return: List of sequences corresponding to the regions.\n",
    "    \"\"\"\n",
    "    sequences = []\n",
    "    for index, row in df.iterrows():\n",
    "        # Extracting the sequence\n",
    "        seq = genome[row['Chromosome']][row['50BP_Start']:row['50BP_End']]\n",
    "        \n",
    "        # Reverse complement if the gene is on the negative strand\n",
    "        if row['Strand'] == '-':\n",
    "            seq = seq.reverse_complement()\n",
    "        seq= seq.upper()\n",
    "        \n",
    "        sequences.append(str(seq))  # Converts Seq object to string\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "224b055c-4387-472b-bbe0-d69222ef0f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Paths\n",
    "genome_path = \"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5d63c31f-2579-40d1-ac7d-a7d68f4f3af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the genome and TSS regions\n",
    "genome = read_genome(genome_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bbe9f12e-cc0c-4a95-aabb-912656e99cfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch sequences\n",
    "sequences = fetch_sequences_for_regions(unique_tss_df, genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6aefb7f2-38a5-4d61-bef6-b8d385c94c3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>50BP_Start</th>\n",
       "      <th>50BP_End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>TATA_Box_Positions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11819</td>\n",
       "      <td>11919</td>\n",
       "      <td>+</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>GTTTAAACGAGATTGCCAGCACCGGGTATCATTCACCATTTTTCTT...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11960</td>\n",
       "      <td>12060</td>\n",
       "      <td>+</td>\n",
       "      <td>12010</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>GGGCCTTTGAGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTG...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17436</td>\n",
       "      <td>17536</td>\n",
       "      <td>-</td>\n",
       "      <td>17436</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>ATCAGACACAGGTACAGCACATAGGCCAGGAGCCAGGGGGTGACGG...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29504</td>\n",
       "      <td>29604</td>\n",
       "      <td>+</td>\n",
       "      <td>29554</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>CGAGTCGCGGGCCTGGGCACGGAACTCACGCTCACTCCGAGCTCCC...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29570</td>\n",
       "      <td>29670</td>\n",
       "      <td>-</td>\n",
       "      <td>29570</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>GAGGGGAAGGCAGCGATGGGACAAAGGACGGAGGTCTAGGAAGAGG...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215293</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209256</td>\n",
       "      <td>57209356</td>\n",
       "      <td>+</td>\n",
       "      <td>57209306</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483079.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-210</td>\n",
       "      <td>CTGTAACAAAGACCCATGTGATGCTGGGGGCAGAGACAGAGGAGAA...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215294</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209837</td>\n",
       "      <td>57209937</td>\n",
       "      <td>+</td>\n",
       "      <td>57209887</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000496301.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-215</td>\n",
       "      <td>ACTAACACCACCCCCACCGCCCCCACCACCACCCCCAGCTCCTGAG...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215295</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210294</td>\n",
       "      <td>57210394</td>\n",
       "      <td>+</td>\n",
       "      <td>57210344</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483286.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-211</td>\n",
       "      <td>CAGTGTAACCCTAGCATCCAGAAGTGGCACAAAACCCCTCTGCTGG...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215296</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210541</td>\n",
       "      <td>57210641</td>\n",
       "      <td>+</td>\n",
       "      <td>57210591</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000464205.6_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>WASH6P-205</td>\n",
       "      <td>TGACTTCCCTCCCAGCAGGCTCTTGGACACAGTAAGCTTCCCCAGC...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215297</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57214397</td>\n",
       "      <td>57214497</td>\n",
       "      <td>-</td>\n",
       "      <td>57214397</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>GGAGAGGAGCCATGCCTAGAGTGGGATGGGCCATTGTTCATCTTCT...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215298 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  50BP_Start  50BP_End Strand       TSS  \\\n",
       "0            chr1       11819     11919      +     11869   \n",
       "1            chr1       11960     12060      +     12010   \n",
       "2            chr1       17436     17536      -     17436   \n",
       "3            chr1       29504     29604      +     29554   \n",
       "4            chr1       29570     29670      -     29570   \n",
       "...           ...         ...       ...    ...       ...   \n",
       "215293       chrY    57209256  57209356      +  57209306   \n",
       "215294       chrY    57209837  57209937      +  57209887   \n",
       "215295       chrY    57210294  57210394      +  57210344   \n",
       "215296       chrY    57210541  57210641      +  57210591   \n",
       "215297       chrY    57214397  57214497      -  57214397   \n",
       "\n",
       "                          GeneID                           gene_type  \\\n",
       "0              ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "1              ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "2              ENSG00000278267.1                               miRNA   \n",
       "3              ENSG00000243485.5                              lncRNA   \n",
       "4              ENSG00000227232.5              unprocessed_pseudogene   \n",
       "...                          ...                                 ...   \n",
       "215293  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215294  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215295  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215296  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "215297   ENSG00000227159.8_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "          gene_name     hgnc_id           havana_gene  \\\n",
       "0           DDX11L1  HGNC:37102  OTTHUMG00000000961.2   \n",
       "1           DDX11L1  HGNC:37102  OTTHUMG00000000961.2   \n",
       "2         MIR6859-1  HGNC:50039                 [nan]   \n",
       "3       MIR1302-2HG  HGNC:52482  OTTHUMG00000000959.2   \n",
       "4            WASH7P  HGNC:38034  OTTHUMG00000000958.1   \n",
       "...             ...         ...                   ...   \n",
       "215293       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215294       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215295       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215296       WASH6P  HGNC:31685  OTTHUMG00000022677.5   \n",
       "215297     DDX11L16  HGNC:37115  OTTHUMG00000022678.1   \n",
       "\n",
       "                  transcript_id                     transcript_type  \\\n",
       "0             ENST00000456328.2                processed_transcript   \n",
       "1             ENST00000450305.2  transcribed_unprocessed_pseudogene   \n",
       "2             ENST00000619216.1                               miRNA   \n",
       "3             ENST00000473358.1                              lncRNA   \n",
       "4             ENST00000488147.1              unprocessed_pseudogene   \n",
       "...                         ...                                 ...   \n",
       "215293  ENST00000483079.6_PAR_Y                     retained_intron   \n",
       "215294  ENST00000496301.6_PAR_Y                     retained_intron   \n",
       "215295  ENST00000483286.6_PAR_Y                     retained_intron   \n",
       "215296  ENST00000464205.6_PAR_Y                processed_transcript   \n",
       "215297  ENST00000507418.6_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "        transcript_name                                           Sequence  \\\n",
       "0           DDX11L1-202  GTTTAAACGAGATTGCCAGCACCGGGTATCATTCACCATTTTTCTT...   \n",
       "1           DDX11L1-201  GGGCCTTTGAGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTG...   \n",
       "2         MIR6859-1-201  ATCAGACACAGGTACAGCACATAGGCCAGGAGCCAGGGGGTGACGG...   \n",
       "3       MIR1302-2HG-202  CGAGTCGCGGGCCTGGGCACGGAACTCACGCTCACTCCGAGCTCCC...   \n",
       "4            WASH7P-201  GAGGGGAAGGCAGCGATGGGACAAAGGACGGAGGTCTAGGAAGAGG...   \n",
       "...                 ...                                                ...   \n",
       "215293       WASH6P-210  CTGTAACAAAGACCCATGTGATGCTGGGGGCAGAGACAGAGGAGAA...   \n",
       "215294       WASH6P-215  ACTAACACCACCCCCACCGCCCCCACCACCACCCCCAGCTCCTGAG...   \n",
       "215295       WASH6P-211  CAGTGTAACCCTAGCATCCAGAAGTGGCACAAAACCCCTCTGCTGG...   \n",
       "215296       WASH6P-205  TGACTTCCCTCCCAGCAGGCTCTTGGACACAGTAAGCTTCCCCAGC...   \n",
       "215297     DDX11L16-201  GGAGAGGAGCCATGCCTAGAGTGGGATGGGCCATTGTTCATCTTCT...   \n",
       "\n",
       "       TATA_Box_Positions  \n",
       "0                      []  \n",
       "1                      []  \n",
       "2                      []  \n",
       "3                      []  \n",
       "4                      []  \n",
       "...                   ...  \n",
       "215293                 []  \n",
       "215294                 []  \n",
       "215295                 []  \n",
       "215296                 []  \n",
       "215297                 []  \n",
       "\n",
       "[215298 rows x 15 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_tss_df['Sequence'] = sequences  # Add sequences to the DataFrame\n",
    "unique_tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "450d9c0c-aafb-448b-b5ac-c4c9c974467a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your existing function\n",
    "def find_tata_box(sequence):\n",
    "    # tata_pattern_plus = r\"TAT[AT][AT][AG]\"  # Simplified TATA box pattern\n",
    "    # tata_pattern_minus= r\"[TC][AT][AT]ATA\"\n",
    "    tata_pattern = r\"TAT[AT][AT][AG]|[TC][AT][AT]ATA\"\n",
    "    for match in re.finditer(tata_pattern, sequence):\n",
    "        yield match.start()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "cd22d3f4-92b4-4155-a7e1-cd2b3fc13e8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming 'unique_tss_df' is your DataFrame with the sequences\n",
    "\n",
    "# Apply the function to each sequence and store the results in a new column\n",
    "unique_tss_df['TATA_Box_Positions'] = unique_tss_df['Sequence'].apply(lambda seq: list(find_tata_box(seq)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "448864cc-abb4-4b5a-b624-c4946d0de937",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the three dataframes based on the conditions\n",
    "# 1. TATA Core Promoter DataFrame\n",
    "tata_core_df = unique_tss_df[unique_tss_df['TATA_Box_Positions'].apply(\n",
    "    lambda positions: any(20 <= pos <= 35 for pos in positions))]\n",
    "\n",
    "# 2. Sequences containing TATA Box but not functioning as TATA core promoter\n",
    "tata_non_functional_df = unique_tss_df[unique_tss_df['TATA_Box_Positions'].apply(\n",
    "    lambda positions: any(pos not in range(20, 36) for pos in positions) if positions else False)]\n",
    "\n",
    "# 3. DataFrame without TATA Box\n",
    "no_tata_df = unique_tss_df[unique_tss_df['TATA_Box_Positions'].apply(len) == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "a03ea461-3ee1-40bb-82b1-e44dc0166c03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>50BP_Start</th>\n",
       "      <th>50BP_End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>TATA_Box_Positions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30217</td>\n",
       "      <td>30317</td>\n",
       "      <td>+</td>\n",
       "      <td>30267</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>AAAGAAAGAAAAGAGGCTTTTTTAAAAAGTTAAATATATAAAAAGA...</td>\n",
       "      <td>[30, 36]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120932</td>\n",
       "      <td>121032</td>\n",
       "      <td>-</td>\n",
       "      <td>120932</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>OTTHUMG00000001096.2</td>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000466430</td>\n",
       "      <td>AATAGATATACAATATATAATATGTAATAGATATATATTATATATA...</td>\n",
       "      <td>[13, 31, 38, 51, 58]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>157887</td>\n",
       "      <td>157987</td>\n",
       "      <td>-</td>\n",
       "      <td>157887</td>\n",
       "      <td>ENSG00000222623.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1100P</td>\n",
       "      <td>HGNC:48063</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>ENST00000410691.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1100P-201</td>\n",
       "      <td>TATTTACTGAGCTGTTACTAGGTGCCTATAAATAGCATAGATTTTT...</td>\n",
       "      <td>[0, 26]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182646</td>\n",
       "      <td>182746</td>\n",
       "      <td>+</td>\n",
       "      <td>182696</td>\n",
       "      <td>ENSG00000279928.2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>HGNC:55080</td>\n",
       "      <td>OTTHUMG00000191962.1</td>\n",
       "      <td>ENST00000624431.2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L17-201</td>\n",
       "      <td>CCCCTGTTGTCTGCATGTAACTTAATACCACAACCAGGCATAGGGG...</td>\n",
       "      <td>[21]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr1</td>\n",
       "      <td>297502</td>\n",
       "      <td>297602</td>\n",
       "      <td>-</td>\n",
       "      <td>297502</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>OTTHUMG00000002552.3</td>\n",
       "      <td>ENST00000424587.7</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>ENST00000424587</td>\n",
       "      <td>ATGTAATAGATATACAATATATAATATATAATAGATATATAATATT...</td>\n",
       "      <td>[17, 24, 35, 45, 58]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215174</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24501505</td>\n",
       "      <td>24501605</td>\n",
       "      <td>-</td>\n",
       "      <td>24501505</td>\n",
       "      <td>ENSG00000232003.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>ZNF736P12Y</td>\n",
       "      <td>HGNC:38431</td>\n",
       "      <td>OTTHUMG00000045107.1</td>\n",
       "      <td>ENST00000443152.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>ZNF736P12Y-201</td>\n",
       "      <td>TTGTCACCTTGTTAATGTGCTTTGTGTAACTAATATTTGTAAAATG...</td>\n",
       "      <td>[29]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215182</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24683607</td>\n",
       "      <td>24683707</td>\n",
       "      <td>+</td>\n",
       "      <td>24683657</td>\n",
       "      <td>ENSG00000226270.3</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>ZNF736P2Y</td>\n",
       "      <td>HGNC:37727</td>\n",
       "      <td>OTTHUMG00000045110.1</td>\n",
       "      <td>ENST00000416040.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>ZNF736P2Y-201</td>\n",
       "      <td>TGAGTCTAATGAGGGAGAGATTTATAGTCCTTTTATTTTCTTAGCT...</td>\n",
       "      <td>[20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215195</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24978317</td>\n",
       "      <td>24978417</td>\n",
       "      <td>-</td>\n",
       "      <td>24978317</td>\n",
       "      <td>ENSG00000278602.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000278602</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>OTTHUMG00000188109.1</td>\n",
       "      <td>ENST00000616864.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENST00000616864</td>\n",
       "      <td>AGAAATTTAAAATATCTATAGTAAATAAATGACATTAATTCTGCAT...</td>\n",
       "      <td>[21]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215241</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25904468</td>\n",
       "      <td>25904568</td>\n",
       "      <td>+</td>\n",
       "      <td>25904518</td>\n",
       "      <td>ENSG00000232914.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>TRAPPC2P4</td>\n",
       "      <td>HGNC:10713</td>\n",
       "      <td>OTTHUMG00000045261.1</td>\n",
       "      <td>ENST00000421750.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>TRAPPC2P4-201</td>\n",
       "      <td>AAAAATATGCACTATTTTTTGCTTCATTAGGAGCCATATATTGAAG...</td>\n",
       "      <td>[34]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215258</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26247334</td>\n",
       "      <td>26247434</td>\n",
       "      <td>+</td>\n",
       "      <td>26247384</td>\n",
       "      <td>ENSG00000251796.1</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>SNORA70</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>ENST00000515987.1</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>SNORA70.11-201</td>\n",
       "      <td>CCTATTTCTCTTCCTAATACTATTTGTTTACACCTATAGTTTTTCT...</td>\n",
       "      <td>[13, 20]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8476 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  50BP_Start  50BP_End Strand       TSS              GeneID  \\\n",
       "5            chr1       30217     30317      +     30267   ENSG00000243485.5   \n",
       "14           chr1      120932    121032      -    120932   ENSG00000238009.6   \n",
       "25           chr1      157887    157987      -    157887   ENSG00000222623.1   \n",
       "31           chr1      182646    182746      +    182696   ENSG00000279928.2   \n",
       "38           chr1      297502    297602      -    297502  ENSG00000228463.10   \n",
       "...           ...         ...       ...    ...       ...                 ...   \n",
       "215174       chrY    24501505  24501605      -  24501505   ENSG00000232003.2   \n",
       "215182       chrY    24683607  24683707      +  24683657   ENSG00000226270.3   \n",
       "215195       chrY    24978317  24978417      -  24978317   ENSG00000278602.1   \n",
       "215241       chrY    25904468  25904568      +  25904518   ENSG00000232914.1   \n",
       "215258       chrY    26247334  26247434      +  26247384   ENSG00000251796.1   \n",
       "\n",
       "                               gene_type        gene_name     hgnc_id  \\\n",
       "5                                 lncRNA      MIR1302-2HG  HGNC:52482   \n",
       "14                                lncRNA  ENSG00000238009       [nan]   \n",
       "25                                 snRNA       RNU6-1100P  HGNC:48063   \n",
       "31                unprocessed_pseudogene         DDX11L17  HGNC:55080   \n",
       "38      transcribed_processed_pseudogene  ENSG00000228463       [nan]   \n",
       "...                                  ...              ...         ...   \n",
       "215174              processed_pseudogene       ZNF736P12Y  HGNC:38431   \n",
       "215182              processed_pseudogene        ZNF736P2Y  HGNC:37727   \n",
       "215195            unprocessed_pseudogene  ENSG00000278602       [nan]   \n",
       "215241            unprocessed_pseudogene        TRAPPC2P4  HGNC:10713   \n",
       "215258                            snoRNA          SNORA70       [nan]   \n",
       "\n",
       "                 havana_gene      transcript_id         transcript_type  \\\n",
       "5       OTTHUMG00000000959.2  ENST00000469289.1                  lncRNA   \n",
       "14      OTTHUMG00000001096.2  ENST00000466430.5                  lncRNA   \n",
       "25                     [nan]  ENST00000410691.1                   snRNA   \n",
       "31      OTTHUMG00000191962.1  ENST00000624431.2  unprocessed_pseudogene   \n",
       "38      OTTHUMG00000002552.3  ENST00000424587.7    processed_transcript   \n",
       "...                      ...                ...                     ...   \n",
       "215174  OTTHUMG00000045107.1  ENST00000443152.1    processed_pseudogene   \n",
       "215182  OTTHUMG00000045110.1  ENST00000416040.1    processed_pseudogene   \n",
       "215195  OTTHUMG00000188109.1  ENST00000616864.1  unprocessed_pseudogene   \n",
       "215241  OTTHUMG00000045261.1  ENST00000421750.1  unprocessed_pseudogene   \n",
       "215258                 [nan]  ENST00000515987.1                  snoRNA   \n",
       "\n",
       "        transcript_name                                           Sequence  \\\n",
       "5       MIR1302-2HG-201  AAAGAAAGAAAAGAGGCTTTTTTAAAAAGTTAAATATATAAAAAGA...   \n",
       "14      ENST00000466430  AATAGATATACAATATATAATATGTAATAGATATATATTATATATA...   \n",
       "25       RNU6-1100P-201  TATTTACTGAGCTGTTACTAGGTGCCTATAAATAGCATAGATTTTT...   \n",
       "31         DDX11L17-201  CCCCTGTTGTCTGCATGTAACTTAATACCACAACCAGGCATAGGGG...   \n",
       "38      ENST00000424587  ATGTAATAGATATACAATATATAATATATAATAGATATATAATATT...   \n",
       "...                 ...                                                ...   \n",
       "215174   ZNF736P12Y-201  TTGTCACCTTGTTAATGTGCTTTGTGTAACTAATATTTGTAAAATG...   \n",
       "215182    ZNF736P2Y-201  TGAGTCTAATGAGGGAGAGATTTATAGTCCTTTTATTTTCTTAGCT...   \n",
       "215195  ENST00000616864  AGAAATTTAAAATATCTATAGTAAATAAATGACATTAATTCTGCAT...   \n",
       "215241    TRAPPC2P4-201  AAAAATATGCACTATTTTTTGCTTCATTAGGAGCCATATATTGAAG...   \n",
       "215258   SNORA70.11-201  CCTATTTCTCTTCCTAATACTATTTGTTTACACCTATAGTTTTTCT...   \n",
       "\n",
       "          TATA_Box_Positions  \n",
       "5                   [30, 36]  \n",
       "14      [13, 31, 38, 51, 58]  \n",
       "25                   [0, 26]  \n",
       "31                      [21]  \n",
       "38      [17, 24, 35, 45, 58]  \n",
       "...                      ...  \n",
       "215174                  [29]  \n",
       "215182                  [20]  \n",
       "215195                  [21]  \n",
       "215241                  [34]  \n",
       "215258              [13, 20]  \n",
       "\n",
       "[8476 rows x 15 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tata_core_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "fe329b71-9204-4f96-9c82-d3f1afaf37c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_core_df.to_csv(\"/data/private/pdutta/DNABERT_data/Core_Prom_new/TATA/raw_positive.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "ede76436-9b31-43bb-891b-f75a27299395",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your existing function\n",
    "def find_tata_box_stringent(sequence):\n",
    "    # tata_pattern_plus = r\"TAT[AT][AT][AG]\"  # Simplified TATA box pattern\n",
    "    # tata_pattern_minus= r\"[TC][AT][AT]ATA\"\n",
    "    tata_pattern = r\"TATA[AT][AT][AG]\"\n",
    "    for match in re.finditer(tata_pattern, sequence):\n",
    "        yield match.start()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "c35f9cc2-587b-483a-af7f-0bd011cb36d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chrX\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "# Initialize a list to hold results\n",
    "results = []\n",
    "#[f'chr{i}' for i in range(1, 23)] +\n",
    "# Iterate through specified chromosomes only\n",
    "for chrom in [f'chr{i}' for i in range(1, 23)]+ ['chrX', 'chrY']:\n",
    "    sequence = genome[chrom]  # Fetch the sequence for the chromosome\n",
    "    length = len(sequence)\n",
    "    print(chrom)\n",
    "    #print(sequence[0:10])\n",
    "\n",
    "    # Break down the chromosome into 100 bp segments and search\n",
    "    for i in range(0, length - 100 + 1, 50):  # 50 bp step for some overlap; adjust as needed\n",
    "        segment = str(sequence[i:i+100]).upper()  # Ensure segment is a string and uppercase\n",
    "        #print(segment)\n",
    "        # Check if segment is a proper string before processing\n",
    "        if isinstance(segment, str):\n",
    "            for pos in find_tata_box_stringent(segment):\n",
    "                if 75 <= pos <= 81:\n",
    "                    results.append({\n",
    "                        'Chromosome': chrom,\n",
    "                        'Start': i,\n",
    "                        'End': i + 100,\n",
    "                        'TATA_Box_Position': pos,\n",
    "                        'Sequence': segment\n",
    "                    })\n",
    "        else:\n",
    "            print(f\"Non-string segment detected at {chrom}:{i}-{i+100}\")\n",
    "\n",
    "# Convert results into a DataFrame\n",
    "tata_box_df = pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "620825d2-b7e1-43bf-be17-68975ca91f9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>TATA_Box_Position</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24300</td>\n",
       "      <td>24400</td>\n",
       "      <td>81</td>\n",
       "      <td>TCTATTGCTTATTTAGGTCTTTCTCTCCTGCCAGAATGTGCGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27350</td>\n",
       "      <td>27450</td>\n",
       "      <td>79</td>\n",
       "      <td>ATTGAGATGTGCCATAAGTGTAAAATATGCACCAAATTTCAAAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43550</td>\n",
       "      <td>43650</td>\n",
       "      <td>78</td>\n",
       "      <td>CCCACAGCTAACATCATAATAGTAAAATCATCACACTGGTAAAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>45650</td>\n",
       "      <td>45750</td>\n",
       "      <td>75</td>\n",
       "      <td>ATGTGTTAATTAACTTGATTTATTCATTCAACAACATACACATATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50700</td>\n",
       "      <td>50800</td>\n",
       "      <td>80</td>\n",
       "      <td>AAAATGCCATTATCCAAAGTGGTTCTCTAGAGATGCTGATGTATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459571</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57150100</td>\n",
       "      <td>57150200</td>\n",
       "      <td>79</td>\n",
       "      <td>ACCCCCTTATCTGGCTGTTCATGTGTTTTCTTTACCATGTCCTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459572</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57151700</td>\n",
       "      <td>57151800</td>\n",
       "      <td>81</td>\n",
       "      <td>TGGGGGGGGCAGTCTGGTGTTACTGAGCCCTTAAACTATTAGGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459573</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57171600</td>\n",
       "      <td>57171700</td>\n",
       "      <td>76</td>\n",
       "      <td>ACTGACCTTATATTGCTAGTCTTTGCTTTTGTACCTCTTTAGGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459574</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57174800</td>\n",
       "      <td>57174900</td>\n",
       "      <td>77</td>\n",
       "      <td>ATACCCGTGAGGAAAATGAAGGACTAGCACATTGGGCCTTAATAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459575</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57175650</td>\n",
       "      <td>57175750</td>\n",
       "      <td>78</td>\n",
       "      <td>CTGAGTGATTTCTTGTAACTTAGAAAAATTAACAGGTTTTATCACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>459576 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End  TATA_Box_Position  \\\n",
       "0            chr1     24300     24400                 81   \n",
       "1            chr1     27350     27450                 79   \n",
       "2            chr1     43550     43650                 78   \n",
       "3            chr1     45650     45750                 75   \n",
       "4            chr1     50700     50800                 80   \n",
       "...           ...       ...       ...                ...   \n",
       "459571       chrY  57150100  57150200                 79   \n",
       "459572       chrY  57151700  57151800                 81   \n",
       "459573       chrY  57171600  57171700                 76   \n",
       "459574       chrY  57174800  57174900                 77   \n",
       "459575       chrY  57175650  57175750                 78   \n",
       "\n",
       "                                                 Sequence  \n",
       "0       TCTATTGCTTATTTAGGTCTTTCTCTCCTGCCAGAATGTGCGCTCC...  \n",
       "1       ATTGAGATGTGCCATAAGTGTAAAATATGCACCAAATTTCAAAGGC...  \n",
       "2       CCCACAGCTAACATCATAATAGTAAAATCATCACACTGGTAAAAAA...  \n",
       "3       ATGTGTTAATTAACTTGATTTATTCATTCAACAACATACACATATA...  \n",
       "4       AAAATGCCATTATCCAAAGTGGTTCTCTAGAGATGCTGATGTATAT...  \n",
       "...                                                   ...  \n",
       "459571  ACCCCCTTATCTGGCTGTTCATGTGTTTTCTTTACCATGTCCTTTA...  \n",
       "459572  TGGGGGGGGCAGTCTGGTGTTACTGAGCCCTTAAACTATTAGGTAT...  \n",
       "459573  ACTGACCTTATATTGCTAGTCTTTGCTTTTGTACCTCTTTAGGGCA...  \n",
       "459574  ATACCCGTGAGGAAAATGAAGGACTAGCACATTGGGCCTTAATAAA...  \n",
       "459575  CTGAGTGATTTCTTGTAACTTAGAAAAATTAACAGGTTTTATCACA...  \n",
       "\n",
       "[459576 rows x 5 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tata_box_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "0f29722d-4eed-49d2-accf-891f04eed5a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert DataFrames to BED format for pybedtools\n",
    "tata_box_bed = pybedtools.BedTool.from_dataframe(tata_box_df[['Chromosome', 'Start', 'End']])\n",
    "tata_core_bed = pybedtools.BedTool.from_dataframe(tata_core_df[['Chromosome', '50BP_Start', '50BP_End']])\n",
    "# Use the intersect function with the -v option to get non-intersecting regions\n",
    "non_intersecting = tata_box_bed.intersect(tata_core_bed, v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "f193d66f-e63d-486c-8540-47a0c87db9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert back to DataFrame\n",
    "non_intersecting_df = non_intersecting.to_dataframe(names=['Chromosome', 'Start', 'End'])\n",
    "\n",
    "# You might want to merge back additional information from tata_box_df if needed\n",
    "tata_box_df_filtered = pd.merge(non_intersecting_df, tata_box_df, on=['Chromosome', 'Start', 'End'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "fe9f5edb-8c4a-4950-8e47-5b8733345a74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>TATA_Box_Position</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24300</td>\n",
       "      <td>24400</td>\n",
       "      <td>81</td>\n",
       "      <td>TCTATTGCTTATTTAGGTCTTTCTCTCCTGCCAGAATGTGCGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27350</td>\n",
       "      <td>27450</td>\n",
       "      <td>79</td>\n",
       "      <td>ATTGAGATGTGCCATAAGTGTAAAATATGCACCAAATTTCAAAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43550</td>\n",
       "      <td>43650</td>\n",
       "      <td>78</td>\n",
       "      <td>CCCACAGCTAACATCATAATAGTAAAATCATCACACTGGTAAAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>45650</td>\n",
       "      <td>45750</td>\n",
       "      <td>75</td>\n",
       "      <td>ATGTGTTAATTAACTTGATTTATTCATTCAACAACATACACATATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50700</td>\n",
       "      <td>50800</td>\n",
       "      <td>80</td>\n",
       "      <td>AAAATGCCATTATCCAAAGTGGTTCTCTAGAGATGCTGATGTATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459193</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57150100</td>\n",
       "      <td>57150200</td>\n",
       "      <td>79</td>\n",
       "      <td>ACCCCCTTATCTGGCTGTTCATGTGTTTTCTTTACCATGTCCTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459194</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57151700</td>\n",
       "      <td>57151800</td>\n",
       "      <td>81</td>\n",
       "      <td>TGGGGGGGGCAGTCTGGTGTTACTGAGCCCTTAAACTATTAGGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459195</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57171600</td>\n",
       "      <td>57171700</td>\n",
       "      <td>76</td>\n",
       "      <td>ACTGACCTTATATTGCTAGTCTTTGCTTTTGTACCTCTTTAGGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459196</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57174800</td>\n",
       "      <td>57174900</td>\n",
       "      <td>77</td>\n",
       "      <td>ATACCCGTGAGGAAAATGAAGGACTAGCACATTGGGCCTTAATAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57175650</td>\n",
       "      <td>57175750</td>\n",
       "      <td>78</td>\n",
       "      <td>CTGAGTGATTTCTTGTAACTTAGAAAAATTAACAGGTTTTATCACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>459198 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End  TATA_Box_Position  \\\n",
       "0            chr1     24300     24400                 81   \n",
       "1            chr1     27350     27450                 79   \n",
       "2            chr1     43550     43650                 78   \n",
       "3            chr1     45650     45750                 75   \n",
       "4            chr1     50700     50800                 80   \n",
       "...           ...       ...       ...                ...   \n",
       "459193       chrY  57150100  57150200                 79   \n",
       "459194       chrY  57151700  57151800                 81   \n",
       "459195       chrY  57171600  57171700                 76   \n",
       "459196       chrY  57174800  57174900                 77   \n",
       "459197       chrY  57175650  57175750                 78   \n",
       "\n",
       "                                                 Sequence  \n",
       "0       TCTATTGCTTATTTAGGTCTTTCTCTCCTGCCAGAATGTGCGCTCC...  \n",
       "1       ATTGAGATGTGCCATAAGTGTAAAATATGCACCAAATTTCAAAGGC...  \n",
       "2       CCCACAGCTAACATCATAATAGTAAAATCATCACACTGGTAAAAAA...  \n",
       "3       ATGTGTTAATTAACTTGATTTATTCATTCAACAACATACACATATA...  \n",
       "4       AAAATGCCATTATCCAAAGTGGTTCTCTAGAGATGCTGATGTATAT...  \n",
       "...                                                   ...  \n",
       "459193  ACCCCCTTATCTGGCTGTTCATGTGTTTTCTTTACCATGTCCTTTA...  \n",
       "459194  TGGGGGGGGCAGTCTGGTGTTACTGAGCCCTTAAACTATTAGGTAT...  \n",
       "459195  ACTGACCTTATATTGCTAGTCTTTGCTTTTGTACCTCTTTAGGGCA...  \n",
       "459196  ATACCCGTGAGGAAAATGAAGGACTAGCACATTGGGCCTTAATAAA...  \n",
       "459197  CTGAGTGATTTCTTGTAACTTAGAAAAATTAACAGGTTTTATCACA...  \n",
       "\n",
       "[459198 rows x 5 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tata_box_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "3d35a6ff-88ce-4c52-9cc0-37aef7a54543",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_box_df_filtered.to_csv(\"/data/private/pdutta/DNABERT_data/Core_Prom_new/TATA/raw_negative_all.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "da105da3-29e7-4fce-b9d1-9da69096fdb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_box_df_filtered_sample = tata_box_df_filtered.sample(n=8500, random_state=1).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "a368d99e-87ad-4a4c-97ab-ca8567b9c5ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>TATA_Box_Position</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>26073800</td>\n",
       "      <td>26073900</td>\n",
       "      <td>76</td>\n",
       "      <td>GACTATACCAAGGCATTTCTACACCTTTCTTTATCATCTGTACATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>7125250</td>\n",
       "      <td>7125350</td>\n",
       "      <td>75</td>\n",
       "      <td>GTTGTCGTTCAGCCTTTCTAACCTGCAGCATCCTCATTTGTGCAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>107362200</td>\n",
       "      <td>107362300</td>\n",
       "      <td>80</td>\n",
       "      <td>TAACAGGCAACGTGCCATGAATACATAAATATGAAAAAGATGCAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>123781250</td>\n",
       "      <td>123781350</td>\n",
       "      <td>79</td>\n",
       "      <td>GAGGCAATGTCTGCTCTTCACTTGATTCCCTTCCATATAAATTGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>149746400</td>\n",
       "      <td>149746500</td>\n",
       "      <td>78</td>\n",
       "      <td>AGGAGTAGTTTCAGTAGAGAGGGGGTGTATTAGTCCTTTCTCACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8495</th>\n",
       "      <td>chr3</td>\n",
       "      <td>105119050</td>\n",
       "      <td>105119150</td>\n",
       "      <td>79</td>\n",
       "      <td>ACACATAAATATAGTCATCCACTGCCTACGGAAATTTTAGTCAATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8496</th>\n",
       "      <td>chr10</td>\n",
       "      <td>10642300</td>\n",
       "      <td>10642400</td>\n",
       "      <td>76</td>\n",
       "      <td>GTCACACCAACCTTCATACTCTCTCTGAATCACTTCTCCAGCACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8497</th>\n",
       "      <td>chr18</td>\n",
       "      <td>1045150</td>\n",
       "      <td>1045250</td>\n",
       "      <td>80</td>\n",
       "      <td>ATACATATATGTAGATTCCATCATATATGGGAACAAAAGTTTCCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8498</th>\n",
       "      <td>chr6</td>\n",
       "      <td>51491650</td>\n",
       "      <td>51491750</td>\n",
       "      <td>79</td>\n",
       "      <td>ACAACATGGGAGAAAGGCAGGCAGAAGCAAGATACCCCATCCACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8499</th>\n",
       "      <td>chr9</td>\n",
       "      <td>108426800</td>\n",
       "      <td>108426900</td>\n",
       "      <td>76</td>\n",
       "      <td>CCTAAAAACAGAAACAGCATTCCACCCAGCAATCCTGTTACTGGAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8500 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome      Start        End  TATA_Box_Position  \\\n",
       "0          chr5   26073800   26073900                 76   \n",
       "1         chr16    7125250    7125350                 75   \n",
       "2          chr8  107362200  107362300                 80   \n",
       "3         chr10  123781250  123781350                 79   \n",
       "4          chr3  149746400  149746500                 78   \n",
       "...         ...        ...        ...                ...   \n",
       "8495       chr3  105119050  105119150                 79   \n",
       "8496      chr10   10642300   10642400                 76   \n",
       "8497      chr18    1045150    1045250                 80   \n",
       "8498       chr6   51491650   51491750                 79   \n",
       "8499       chr9  108426800  108426900                 76   \n",
       "\n",
       "                                               Sequence  \n",
       "0     GACTATACCAAGGCATTTCTACACCTTTCTTTATCATCTGTACATT...  \n",
       "1     GTTGTCGTTCAGCCTTTCTAACCTGCAGCATCCTCATTTGTGCAGT...  \n",
       "2     TAACAGGCAACGTGCCATGAATACATAAATATGAAAAAGATGCAGT...  \n",
       "3     GAGGCAATGTCTGCTCTTCACTTGATTCCCTTCCATATAAATTGAC...  \n",
       "4     AGGAGTAGTTTCAGTAGAGAGGGGGTGTATTAGTCCTTTCTCACAC...  \n",
       "...                                                 ...  \n",
       "8495  ACACATAAATATAGTCATCCACTGCCTACGGAAATTTTAGTCAATG...  \n",
       "8496  GTCACACCAACCTTCATACTCTCTCTGAATCACTTCTCCAGCACAG...  \n",
       "8497  ATACATATATGTAGATTCCATCATATATGGGAACAAAAGTTTCCAT...  \n",
       "8498  ACAACATGGGAGAAAGGCAGGCAGAAGCAAGATACCCCATCCACTT...  \n",
       "8499  CCTAAAAACAGAAACAGCATTCCACCCAGCAATCCTGTTACTGGAT...  \n",
       "\n",
       "[8500 rows x 5 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tata_box_df_filtered_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "8269260f-ccba-4d08-aa85-ce0ec59f9dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_box_df_filtered_sample.to_csv(\"/data/private/pdutta/DNABERT_data/Core_Prom_new/TATA/raw_negative.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6d21d321-a58c-493f-8d54-0de0fa4a6db5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4320</th>\n",
       "      <td>GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4321</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4322</th>\n",
       "      <td>CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4324</th>\n",
       "      <td>CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236975 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence\n",
       "0     GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...\n",
       "1     GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...\n",
       "2     CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...\n",
       "3     TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...\n",
       "4     TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...\n",
       "...                                                 ...\n",
       "4320  GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...\n",
       "4321  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...\n",
       "4322  CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...\n",
       "4323  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...\n",
       "4324  CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...\n",
       "\n",
       "[236975 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "86981274-16dc-4444-ae56-df2b19be889c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos['Label']=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "53e66247-0acf-41d9-b6fc-02bba72480f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236970</th>\n",
       "      <td>GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236971</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236972</th>\n",
       "      <td>CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236973</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236974</th>\n",
       "      <td>CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236975 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Sequence  Label\n",
       "0       GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...      1\n",
       "1       GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...      1\n",
       "2       CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...      1\n",
       "3       TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...      1\n",
       "4       TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...      1\n",
       "...                                                   ...    ...\n",
       "236970  GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...      1\n",
       "236971  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...      1\n",
       "236972  CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...      1\n",
       "236973  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...      1\n",
       "236974  CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...      1\n",
       "\n",
       "[236975 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos = df_pos.reset_index(drop=True)\n",
    "df_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0fd94088-8542-4675-850a-5850e95cb0ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos.to_csv(\"/data/projects/DNABERT_data/Core_promoters/Core_promoter_regions/positive_set.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3c073550-550d-4db9-a474-8bdf50ab0850",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d27f37d9-a1b9-45ab-a2f4-637f59be2453",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
